rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/370_join_hits.yml

```diff
@@ -0,0 +1,237 @@
+
+  - do:
+      index:
+        index: machines
+        id: '192.168.1.2'
+        body: { name: 'The second machine', os: 'Macos', sensitivity: confidential }
+
+  - do:
+      index:
+        index: machines
+        id: '192.168.1.3'
+        body: { name: 'The third machine', os: 'Macos', sensitivity: confidential }
+
+  - do:
+      index:
+        index: machines
+        id: '192.168.1.4'
+        body: { name: 'The forth machine', os: 'Windows', sensitivity: private }
+
+  - do:
+      index:
+        index: machines
+        id: '192.168.1.4'
+        body: { name: 'u2', os: 'Windows', sensitivity: private }
+
+  - do:
+      indices.create:
+        index: logs
+        body:
+          mappings:
+            properties:
+              ip:
+                type: keyword
+              user:
+                type: keyword
+              message:
+                type: text
+              timestamp:
+                type: date
+                format: 'yyyy-MM-dd HH:mm:ss'
+  - do:
+      index:
+        index: logs
+        body: { ip: '192.168.1.1', user: u1, message: 'Logged in', timestamp: '2021-02-11 01:14:01' }
+
+  - do:
+      index:
+        index: logs
+        body: { ip: '192.168.1.1', user: u1, message: 'delete [workspace]', timestamp: '2021-04-11 08:15:02' }
+
+  - do:
+      index:
+        index: logs
+        body: { ip: '192.168.1.1', user: u1, message: 'accessed [documents]', timestamp: '2021-04-11 08:16:01' }
+
+  - do:
+      index:
+        index: logs
+        body: { ip: '192.168.1.1', user: u1, message: 'deleted [documents]', timestamp: '2021-04-11 08:21:10' }
+
+  - do:
+      index:
+        index: logs
+        body: { ip: '192.168.1.1', user: u2, message: 'accessed [workspace]', timestamp: '2021-06-11 11:10:23' }
+
+  - do:
+      index:
+        index: logs
+        body: { ip: '192.168.1.2', user: u2, message: 'copied files in [documents]', timestamp: '2021-07-11 10:20:12' }
+
+  - do:
+      index:
+        index: logs
+        body: { ip: '192.168.1.4', user: u3, message: 'accessed [workspace]', timestamp: '2021-08-11 09:10:12' }
+
+
+  - do:
+      indices.refresh:
+        index: "users,machines,logs"
+
+---
+"Join fields":
+  - skip:
+      version: " - 7.99.99"
+      reason: "Join hits are introduced in 8.0"
+  - do:
+      search:
+        index: logs
+        body:
+          query:
+            match:
+              message: 'documents'
+          _source: false
+          fields:
+            - user
+            - ip
+            - message
+            - timestamp
+          join_hits:
+            - name: host
+              index: machines
+              match_field: ip
+
+            - name: user
+              index: users
+              match_field: user
+              fields: [ name, department ]
+
+            - name: adjacents
+              index: logs
+              query:
+                bool:
+                  must:
+                    - term:
```

Comment:
Should we also support queries like this one?

Reply:
Also we need to think about what our ingestion tools need to do to ensure that the data can be joined up (i.e. order of indexing across different indices, a property we don't have e.g. for bulk requests)